jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,275 @@
+        not be added.
+
+        The proxy implementation should also have a way to associate the fact that a
+        route came from JupyterHub.
+        """
+
+        self.log.info("Adding route for %s to %s.", routespec, target)
+
+        (
+            backend_alias,
+            backend_url_path,
+            frontend_alias,
+            frontend_rule_path,
+        ) = traefik_utils.generate_route_keys(self, target, routespec, separator=".")
+
+        data = json.dumps(data)
+        rule = traefik_utils.generate_rule(routespec)
+
+        await self.mutex.acquire()
+        try:
+            self.routes_cache[routespec] = {
+                "backend": [
+                    "\t[backends." + backend_alias + "]\n",
+                    "\t\t[" + backend_url_path + "]\n",
+                    "\t\t\turl = " + '"' + target + '"\n',
+                    "\t\t\tweight = " + "1\n",
+                ],
+                "frontend": [
+                    "\t[frontends." + frontend_alias + "]\n",
+                    "\t\tbackend = " + '"' + backend_alias + '"\n',
+                    "\t\tpassHostHeader = true\n",
+                    "\t\t[" + frontend_rule_path + "]\n",
+                    "\t\t\trule = " + '"' + rule + '"\n',
+                    "\t\t\tdata = " + "'" + data + "'\n",
+                ],
+                "data": data,
+                "target": target,
+            }
+            self._persist_routes()
+            # self._update_config_file()
+        finally:
+            self.mutex.release()
+
+        try:
+            # Check if traefik was launched
+            pid = self.traefik_process.pid
+            await self._wait_for_route(target, provider="file")
+        except AttributeError:
+            self.log.error(
+                "You cannot add routes if the proxy isn't running! Please start the proxy: proxy.start()"
+            )
+            raise
+
+    async def delete_route(self, routespec):
+        """Delete a route with a given routespec if it exists.
+
+        **Subclasses must define this method**
+        """
+        await self.mutex.acquire()
+        try:
+            del self.routes_cache[routespec]
+            self._persist_routes()
+            # self._update_config_file()
+        finally:
+            self.mutex.release()
+
+    async def get_all_routes(self):
+        """Fetch and return all the routes associated by JupyterHub from the
+        proxy.
+
+        **Subclasses must define this method**
+
+        Should return a dictionary of routes, where the keys are
+        routespecs and each value is a dict of the form::
+
+          {
+            'routespec': the route specification ([host]/path/)
+            'target': the target host URL (proto://host) for this route
+            'data': the attached data dict for this route (as specified in add_route)
+          }
+        """
+        all_routes = {}
+        await self.mutex.acquire()
```

Comment:
This can be easier to keep track of with an [async context mananger](https://docs.python.org/3/glossary.html#term-asynchronous-context-manager), which the mutex is:

```python
async with self.mutex: